### PR TITLE
Added ability to estimate token count for tools

### PR DESCRIPTION
**Pull request by ghost**
Also refactored MessageWindowChatMemory and TokenWindowChatMemory

### CODE REVIEW

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/MessageWindowChatMemory.java [hunk 1]
```diff
@@ -2,104 +2,108 @@
...
         }
 
-        log.debug("Current message count: {}", getCurrentNumberOfMessages());
+        log.debug("Current message count: {}", currentMessageCount());
     }
 
-    private int getCurrentNumberOfMessages() {
-        return maybeSystemMessage.map(m -> 1).orElse(0) + previousMessages.size();
+    private int currentMessageCount() {
+        return messages().size();
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/MessageWindowChatMemory.java:71]: bug: should take system message into account

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/MessageWindowChatMemory.java [hunk 2]
```diff
@@ -2,104 +2,108 @@
...
         private Integer capacity;
-        private LinkedList<ChatMessage> previousMessages = new LinkedList<>();
+        private SystemMessage systemMessage;
+        private LinkedList<ChatMessage> messages = new LinkedList<>();
 
-        public Builder systemMessage(SystemMessage systemMessage) {
-            this.maybeSystemMessage = Optional.ofNullable(systemMessage);
+        public Builder capacity(Integer capacity) {
+            if (capacity < 1) {
+                throw illegalArgument("Capacity should be greater than 1");
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/MessageWindowChatMemory.java]: move to ctor

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/MessageWindowChatMemory.java [hunk 3]
```diff
@@ -2,104 +2,108 @@
...
         public Builder systemMessage(String systemMessage) {
-            if (systemMessage == null) {
-                this.maybeSystemMessage = Optional.empty();
-                return this;
-            }
-
             return systemMessage(SystemMessage.from(systemMessage));
         }
 
-        public Builder capacityInMessages(Integer capacityInMessages) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/MessageWindowChatMemory.java:92]: capacityMessages?
capacity?

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/MessageWindowChatMemory.java [hunk 4]
```diff
@@ -108,11 +112,7 @@ public MessageWindowChatMemory build() {
         }
     }
 
-    public static Builder builder() {
-        return new Builder();
-    }
-
-    public static MessageWindowChatMemory withCapacity(int n) {
-        return builder().capacityInMessages(n).build();
+    public static MessageWindowChatMemory withCapacity(int capacity) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/MessageWindowChatMemory.java]: capacity -> capacityMessages?
add javadoc

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/TokenWindowChatMemory.java [hunk 5]
```diff
@@ -11,112 +10,107 @@
...
+                    tokenCountOfOldestMessage, oldestMessage);
+            currentTokenCount -= tokenCountOfOldestMessage;
         }
 
-        log.debug("Current token count: {}", getCurrentTokenCount());
+        log.debug("Current token count: {}", currentTokenCount());
+    }
+
+    private int currentTokenCount() {
+        return tokenizer.estimateTokenCountInMessages(messages());
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/TokenWindowChatMemory.java:76]: bug: check system message as well

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/TokenWindowChatMemory.java [hunk 6]
```diff
@@ -11,112 +10,107 @@
...
         private Tokenizer tokenizer;
-        private Optional<SystemMessage> maybeSystemMessage = Optional.empty();
-        private Integer capacityInTokens = DEFAULT_CAPACITY_IN_TOKENS;
-        private LinkedList<ChatMessage> previousMessages = new LinkedList<>();
+        private SystemMessage systemMessage;
+        private LinkedList<ChatMessage> messages = new LinkedList<>();
 
-        public Builder tokenizer(Tokenizer tokenizer) {
+        public Builder capacity(Integer capacity, Tokenizer tokenizer) {
+            if (capacity < 1) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/TokenWindowChatMemory.java]: move to ctor

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/TokenWindowChatMemory.java [hunk 7]
```diff
@@ -11,112 +10,107 @@
...
+        private Integer capacity;
         private Tokenizer tokenizer;
-        private Optional<SystemMessage> maybeSystemMessage = Optional.empty();
-        private Integer capacityInTokens = DEFAULT_CAPACITY_IN_TOKENS;
-        private LinkedList<ChatMessage> previousMessages = new LinkedList<>();
+        private SystemMessage systemMessage;
+        private LinkedList<ChatMessage> messages = new LinkedList<>();
 
-        public Builder tokenizer(Tokenizer tokenizer) {
+        public Builder capacity(Integer capacity, Tokenizer tokenizer) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/TokenWindowChatMemory.java]: capacityTokens?

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/TokenWindowChatMemory.java [hunk 8]
```diff
@@ -125,7 +119,7 @@ public TokenWindowChatMemory build() {
         }
     }
 
-    public static Builder builder() {
-        return new Builder();
+    public static TokenWindowChatMemory withCapacity(int capacity, Tokenizer tokenizer) {
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/TokenWindowChatMemory.java]: capacityTokens?

File: langchain4j/src/main/java/dev/langchain4j/model/openai/OpenAiStreamingLanguageModel.java [hunk 9]
```diff
@@ -27,7 +28,6 @@ public OpenAiStreamingLanguageModel(String apiKey,
                                         Boolean logRequests,
                                         Boolean logResponses) {
 
-
         modelName = modelName == null ? TEXT_DAVINCI_003 : modelName;
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/model/openai/OpenAiStreamingLanguageModel.java:37]: add javadoc to all OpenAI language models with note to prefer chat models

File: langchain4j/src/main/java/dev/langchain4j/model/openai/OpenAiTokenizer.java [hunk 10]
```diff
@@ -56,31 +58,61 @@ public int countTokens(ChatMessage message) {
...
+                            tokenCount += 3; // found experimentally while playing with OpenAI API
+                            tokenCount += estimateTokenCountInText(enumValue.toString());
+                        }
+                    }
+                }
+            }
+            tokenCount += 12; // found experimentally while playing with OpenAI API
+        }
+        tokenCount += 12; // found experimentally while playing with OpenAI API
+        return tokenCount;
```
Comment (langchain4j) [inline langchain4j/src/main/java/dev/langchain4j/model/openai/OpenAiTokenizer.java:113]: count ToolParameters.required as well

File: langchain4j/src/test/java/dev/langchain4j/memory/chat/MessageWindowChatMemoryTest.java [hunk 11]
```diff
@@ -119,21 +117,24 @@ void should_keep_specified_number_of_messages_in_chat_history_without_system_mes
     }
 
     @Test
-    void should_load_previous_messages_with_message_restriction() {
+    void should_load_messages_with_message_restriction() {
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/memory/chat/MessageWindowChatMemoryTest.java]: clearer name

File: langchain4j/src/test/java/dev/langchain4j/memory/chat/MessageWindowChatMemoryTest.java [hunk 12]
```diff
@@ -119,21 +117,24 @@ void should_keep_specified_number_of_messages_in_chat_history_without_system_mes
...
 
         List<ChatMessage> previousMessages = asList(
-                userMessageWith(10),
-                aiMessageWith(10),
-                userMessageWith(10),
-                aiMessageWith(10)
+                userMessage("first"),
+                aiMessage("second"),
+                userMessage("third"),
+                aiMessage("fourth")
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/memory/chat/MessageWindowChatMemoryTest.java]: add fifth

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 13]
```diff
@@ -18,9 +14,19 @@
 import static java.util.Arrays.stream;
 import static java.util.stream.Collectors.toList;
 
-class ToolSpecifications {
+public class ToolSpecifications {
+
+    public static List<ToolSpecification> toolSpecificationsFrom(Object objectWithTools) {
+        List<ToolSpecification> toolSpecifications = new ArrayList<>();
```
Comment (kuraleta) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java]: `    return stream(objectWithTools.getClass().getDeclaredMethods())
                .filter(method -> method.isAnnotationPresent(Tool.class))
                .map(ToolSpecifications::toolSpecificationFrom)
                .collect(toList());`

File: langchain4j/src/test/java/dev/langchain4j/model/openai/OpenAiTokenizerTest.java [hunk 14]
```diff
@@ -54,6 +56,34 @@ static Stream<Arguments> should_count_tokens_for_messages() {
...
+    }
+
+    enum TemperatureUnit {
+        F, C
+    }
+
+    @Test
+    void should_count_tokens_in_tools() {
+        int tokenCount = tokenizer.estimateTokenCountInTools(new Tools());
+        assertThat(tokenCount).isEqualTo(93);
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/model/openai/OpenAiTokenizerTest.java]: where from 93?

File: langchain4j/src/test/java/dev/langchain4j/service/AiServicesIT.java [hunk 15]
```diff
@@ -617,7 +622,71 @@ void should_keep_separate_chat_memory_for_each_user() {
...
+
+        verify(chatLanguageModel).sendMessages(
+                singletonList(messages.get(0)),
+                singletonList(ToolSpecification.builder()
+                        .name("squareRoot")
+                        .description("calculates the square root of the provided number")
+                        .addParameter("arg0", NUMBER, JsonSchemaProperty.description("number to operate on"))
+                        .build())
+        );
+        verify(chatLanguageModel).sendMessages(asList(messages.get(0), messages.get(1), messages.get(2)));
```
Comment (langchain4j) [inline langchain4j/src/test/java/dev/langchain4j/service/AiServicesIT.java:693]: add explicit comment that tools are not sent this time

**Review by kuraleta — APPROVED** [review]
good job!

**Pull request merged**